Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,334 @@
+      return;
+    }
+    set_associated_model_value(m_current, false);
+    m_is_blocked = true;
+    set_current(value);
+    m_is_blocked = false;
+  }
+
+  template <typename T>
+  AssociativeValueModel<boost::optional<T>>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template <typename T>
+  void AssociativeValueModel<boost::optional<T>>::associate(
+      const std::shared_ptr<BooleanModel>& model, const T& value) {
+    if(m_models.find(value) != m_models.end()) {
+      return;
+    }
+    m_models[value] = {model, model->connect_current_signal(
+      [=] (auto is_selected) { on_current(value, is_selected); })};
+    if(model->get_current() == true) {
+      if(m_current) {
+        set_associated_model_value(value, false);
+      } else {
+        set_current(value);
+      }
+    }
+  }
+
+  template <typename T>
+  std::shared_ptr<BooleanModel>
+      AssociativeValueModel<boost::optional<T>>::make_association(
+      const T& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template <typename T>
+  std::shared_ptr<BooleanModel> 
+      AssociativeValueModel<boost::optional<T>>::find(const T& value) const {
+    if(auto iterator = m_models.find(value); iterator != m_models.end()) {
+      return iterator->second.m_model;
+    }
+    return nullptr;
+  }
+
+  template <typename T>
+  QValidator::State
+      AssociativeValueModel<boost::optional<T>>::get_state() const {
+    if(!m_models.empty()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template <typename T>
+  const boost::optional<T>&
+      AssociativeValueModel<boost::optional<T>>::get_current() const {
+    return m_current;
+  }
+
+  template <typename T>
+  QValidator::State AssociativeValueModel<boost::optional<T>>::set_current(
+      const boost::optional<T>& value) {
+    if(value && m_models.find(*value) == m_models.end()) {
+      return QValidator::Invalid;
+    }
+    if(m_current) {
+      set_associated_model_value(*m_current, false);
+    }
+    m_current = value;
+    m_current_signal(m_current);
+    return QValidator::Acceptable;
+  }
+
+  template <typename T>
+  boost::signals2::connection
+      AssociativeValueModel<boost::optional<T>>::connect_current_signal(
+      const typename ValueModel<boost::optional<T>>::CurrentSignal::slot_type& slot) const {
+    return m_current_signal.connect(slot);
+  }
+
+  template <typename T>
+  void AssociativeValueModel<boost::optional<T>>::set_associated_model_value(
+      const T& value, bool model_value) {
+    if(auto iterator = m_models.find(value); iterator != m_models.end()) {
+      auto blocker = boost::signals2::shared_connection_block(
+        iterator->second.m_connection);
+      iterator->second.m_model->set_current(model_value);
+    }
+  }
+
+  template <typename T>
+  void AssociativeValueModel<boost::optional<T>>::on_current(const T& value,
```

Comment:
Not a big fan of all this duplicate work for this specialization. My hope was that ```AssociativeValueModel<T>``` would be implemented in terms of ```AssociativeValueModel<optional<T>>```, or at the very least there's be some unification between the two. There is way too much duplication here meaning that all of the issues with the ```T``` version have to get fixed with the ```optional<T>``` as well and retested.

Reply:
With the removal of the m_current variable I should be able to share the members of both implementations, and could implement it using a method like this: https://stackoverflow.com/a/41284020/11234391

If the T version was implemented using the optional<T>, wouldn't it have the same problem with having to test it twice, because I'd have to test both versions? I'm having a hard time visualizing what an implementation like that would look like to simplify the current implementation.

Reply:
Using tag dispatch would be a great approach and it's a very sound way to solve problems like this, if you can do it that way then go for it. The second best way would be using ```constexpr if```s, the third best way would be class specializations or ```SFINAE```.

You would still be testing both versions, but you'd be testing different things about them. With the way this is implemented you have to test many of the exact same things twice.

Reply:
After reviewing it, I can't see how a tag dispatch implementation would work because it requires a 'nested' template type where the regular type is T, but the optional type is optional<T>. In the class, any reference to T would also include the optional type.

For implementing the T typed model in terms of the optional<T> typed model, was the idea to have an instance of the optional type model wrapped in the regular type model?